ghcide/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -86,7 +125,43 @@ instance NFData ShowDiagnostic where
 --   along with the related source location so that we can display the error
 --   on either the console or in the IDE at the right source location.
 --
-type FileDiagnostic = (NormalizedFilePath, ShowDiagnostic, Diagnostic)
+data FileDiagnostic = FileDiagnostic
+  { fdFilePath :: NormalizedFilePath
+  , fdShouldShowDiagnostic :: ShowDiagnostic
+  , fdLspDiagnostic :: Diagnostic
+  , fdStructuredMessage :: StructuredMessage
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+instance NFData FileDiagnostic
+
+data StructuredMessage
+  = NoStructuredMessage
+  | SomeStructuredMessage (MsgEnvelope GhcMessage)
+  deriving (Generic)
+
+instance Show StructuredMessage where
+  show NoStructuredMessage = "NoStructuredMessage"
+  show SomeStructuredMessage {} = "SomeStructuredMessage"
+
+instance Eq StructuredMessage where
+  (==) NoStructuredMessage NoStructuredMessage = True
+  (==) SomeStructuredMessage {} SomeStructuredMessage {} = True
+  (==) _ _ = False
+
+instance Ord StructuredMessage where
+  compare NoStructuredMessage NoStructuredMessage = EQ
+  compare SomeStructuredMessage {} SomeStructuredMessage {} = EQ
+  compare NoStructuredMessage SomeStructuredMessage {} = GT
+  compare SomeStructuredMessage {} NoStructuredMessage = LT
+
+instance NFData StructuredMessage where
+  rnf NoStructuredMessage = ()
+  rnf SomeStructuredMessage {} = ()
+
+modifyFdLspDiagnostic :: (Diagnostic -> Diagnostic) -> FileDiagnostic -> FileDiagnostic
```

Comment:
We use lenses a lot, this seems like a really obvious place for a lens! Just generate lenses for the `FileDiagnostic` type.

Reply:
Fixed in 12b364ae